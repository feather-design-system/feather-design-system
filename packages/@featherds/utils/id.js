const getSafeId = function (str) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
