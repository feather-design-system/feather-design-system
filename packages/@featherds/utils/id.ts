const getSafeId = function (str: string) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
