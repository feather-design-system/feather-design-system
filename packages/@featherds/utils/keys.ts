const KEYCODES = {
  PAGEUP: 33,
  PAGEDOWN: 34,
  UP: 38,
  DOWN: 40,
  LEFT: 37,
  RIGHT: 39,
  HOME: 36,
  END: 35,
  ENTER: 13,
  SPACE: 32,
  ESCAPE: 27,
  BACKSPACE: 8,
  DELETE: 46,
};

export { KEYCODES };
