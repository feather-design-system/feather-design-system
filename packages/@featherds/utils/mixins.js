var translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (key) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};
export { translateKey };
