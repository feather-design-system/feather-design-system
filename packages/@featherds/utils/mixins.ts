const translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (this: IComponent, key: string) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};

interface IComponent {
  translationKey: string;
  $t: (s: string) => void;
}
export { translateKey };
