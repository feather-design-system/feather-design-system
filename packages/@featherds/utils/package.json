{
  "name": "@featherds/utils",
  "version": "0.2.10",
  "publishConfig": {
    "access": "public"
  },
  "author": "NantHealth",
  "scripts": {},
  "keywords": [
    "vue",
    "FeatherUtils"
  ],
  "types": "./src/index.d.ts"
}
