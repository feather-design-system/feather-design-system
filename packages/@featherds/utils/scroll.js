var toView = function (el, parent) {
  if (!el || !parent) {
    return;
  }
  var elHeight = el.getBoundingClientRect().height;
  const compStyle = getComputedStyle(el);
  elHeight += parseInt(compStyle["margin-top"], 10);
  elHeight += parseInt(compStyle["margin-bottom"], 10);
  parent.scrollTop =
    el.offsetTop - parent.getBoundingClientRect().height + elHeight;
};

export { toView };
