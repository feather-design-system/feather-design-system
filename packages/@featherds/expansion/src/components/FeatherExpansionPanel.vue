<template>
  <div class="feather-expansion">
    <div :aria-level="level" role="heading">
      <a
        href="#"
        class="feather-expansion-header-button"
        role="button"
        :class="{ expanded: expanded, disabled: disabled }"
        :id="headerId"
        :aria-controls="contentId"
        :aria-expanded="expanded ? 'true' : 'false'"
        :aria-disabled="disabled ? 'true' : 'false'"
        :aria-busy="loading ? 'true' : 'false'"
        @click.prevent="handleClick"
        data-ref-id="feather-expansion-header-button"
      >
        <span class="feather-expansion-header-button-text">
          <slot name="title"> {{ title }}</slot></span
        >

        <FeatherIcon
          class="feather-expansion-header-button-icon"
          :icon="downIcon"
          :class="{ rotated: expanded, disabled: disabled }"
        />
      </a>
    </div>

    <transition
      name="expand"
      @enter="enter"
      @after-enter="afterEnter"
      @leave="leave"
    >
      <div
        :id="contentId"
        :aria-labelledby="headerId"
        v-if="expanded"
        role="region"
        data-ref-id="feather-expansion-content"
      >
        <div v-if="loading" class="panel-content">
          <FeatherSpinner />
        </div>
        <div v-else class="panel-content">
          <slot />
        </div>
      </div>
    </transition>
  </div>
</template>

<script lang="ts">
import { FeatherIcon } from "@featherds/icon";
import { FeatherSpinner } from "@featherds/progress";
import DownArrow from "@featherds/icon/navigation/ExpandMore";
import { getSafeId } from "@featherds/utils/id";
import { defineComponent } from "vue";
export const props = {
  modelValue: {
    type: Boolean,
    default: false,
    required: false,
  },
  level: {
    type: Number,
    default: 4,
    required: false,
  },
  title: {
    type: String,
    required: false,
  },
  disabled: {
    type: Boolean,
    default: false,
    required: false,
  },
  loading: {
    type: Boolean,
    default: false,
    required: false,
  },
};
export const emits = {
  "update:modelValue": (v: boolean) => true,
};
export default defineComponent({
  name: "FeatherExpansionPanel",
  model: {
    prop: "modelValue",
    event: "update:modelValue",
  },
  emits,
  props,
  watch: {
    modelValue: {
      immediate: true,
      handler(v: boolean) {
        this.internalState = v;
      },
    },
  },
  data() {
    return {
      internalState: this.modelValue,
    };
  },
  computed: {
    expanded() {
      if (this.disabled) {
        return false;
      }
      return this.internalState;
    },
    downIcon() {
      return DownArrow;
    },
    contentId() {
      return getSafeId("feather-expansion-panel");
    },
    headerId() {
      return getSafeId("feather-expansion-header");
    },
  },
  methods: {
    handleClick() {
      if (this.disabled) {
        return;
      }
      this.internalState = !this.internalState;
      this.$emit("update:modelValue", this.internalState);
    },
    enter(el: Element) {
      const element = el as HTMLElement;
      const width = getComputedStyle(element).width;

      element.style.width = width;
      element.style.position = "absolute";
      element.style.visibility = "hidden";
      element.style.height = "auto";

      const height = getComputedStyle(element).height;

      element.style.width = "";
      element.style.position = "";
      element.style.visibility = "";
      element.style.height = "0";

      // Force repaint to make sure the
      // animation is triggered correctly.
      getComputedStyle(element).height;

      // Trigger the animation.
      // We use `setTimeout` because we need
      // to make sure the browser has finished
      // painting after setting the `height`
      // to `0` in the line above.
      setTimeout(() => {
        element.style.height = height;
      });
    },
    afterEnter(el: Element) {
      const element = el as HTMLElement;
      element.style.height = "auto";
    },
    leave(el: Element) {
      const element = el as HTMLElement;
      const height = getComputedStyle(element).height;

      element.style.height = height;

      // Force repaint to make sure the
      // animation is triggered correctly.
      getComputedStyle(element).height;

      setTimeout(() => {
        element.style.height = "0";
      });
    },
  },
  components: {
    FeatherIcon,
    FeatherSpinner,
  },
});
</script>

<style lang="scss" scoped>
@import "@featherds/styles/themes/variables";
@import "@featherds/styles/mixins/typography";
@import "@featherds/styles/mixins/elevation";

.feather-expansion {
  width: 100%;
  @include elevation(2);
  [role="region"] {
    will-change: height;
    .panel-content {
      padding: 0 1.5rem 2rem 1.5rem;
    }
  }

  .feather-expansion-header-button {
    display: flex;
    align-items: center;
    background: transparent;
    border: 0;
    padding: 0 1.5rem;
    position: relative;
    height: 3rem;
    text-align: left;
    width: 100%;
    transition: height 280ms ease-in-out;
    cursor: pointer;
    @include body-small();
    color: var($primary-text-on-surface);
    &:hover {
      text-decoration: none;
    }
    &.expanded {
      height: 4rem;
    }
    &.disabled {
      color: var($disabled-text-on-surface);
      cursor: default;
    }
  }
  .feather-expansion-header-button-text {
    flex: 1;
    display: block;
  }
  .feather-expansion-header-button-icon {
    flex: none;
    display: flex;
    align-items: center;
    justify-items: center;
    height: 1.125rem;
    width: 1.125rem;
    font-size: 1.125rem;
    color: var($secondary-text-on-surface);
    &.disabled {
      color: var($disabled-text-on-surface);
      cursor: default;
    }
  }
}

.feather-expansion-header-button-icon {
  transition: transform 280ms ease-in-out;
  transform-origin: center center;
  &.rotated {
    transform: rotate(180deg);
  }
}
.expand-enter-active,
.expand-leave-active {
  transition: height 250ms ease-in-out;
  overflow: hidden;
}

.expand-enter-from,
.expand-leave-to {
  height: 0;
}
</style>
