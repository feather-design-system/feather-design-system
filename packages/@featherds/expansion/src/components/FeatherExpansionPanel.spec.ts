import FeatherExpansionPanel from "./FeatherExpansionPanel.vue";
import { mount } from "@vue/test-utils";

import axe from "@featherds/utils/test/axe";
import { getCalls } from "@featherds/utils/test/calls";
import * as id from "@featherds/utils/id";
jest.spyOn(id, "getSafeId").mockImplementation((x) => x);

const getWrapper = function () {
  const options = {
    slots,
    props: {
      title,
    },
  };
  return mount(FeatherExpansionPanel, options);
};
const title = "TITLE";
const slots = {
  default: "<p>I AM TEXT</p>",
};
describe("FeatherExpansionPanel.vue", () => {
  it("should toggle expanded value when button is clicked", async () => {
    const wrapper = getWrapper();
    const button = wrapper.find(
      "[data-ref-id='feather-expansion-header-button']"
    );
    await button.trigger("click");
    expect(button.attributes("aria-expanded")).toBe("true");
    expect(getCalls<[boolean]>(wrapper, "update:modelValue")[0][0]).toBe(true);
    await button.trigger("click");
    expect(button.attributes("aria-expanded")).toBe("false");
    expect(getCalls<[boolean]>(wrapper, "update:modelValue")[1][0]).toBe(false);
  });

  it("should not toggle expanded when disabled", async () => {
    const wrapper = getWrapper();
    await wrapper.setProps({ disabled: true });
    const button = wrapper.find(
      "[data-ref-id='feather-expansion-header-button']"
    );
    await button.trigger("click");
    expect(button.attributes("aria-expanded")).toBe("false");
    expect(wrapper.emitted("update:modelValue")).toBeUndefined();
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should render content when expanded", async () => {
    const wrapper = getWrapper();
    await wrapper.setProps({ modelValue: true });

    expect(
      wrapper.find("[data-ref-id='feather-expansion-content']").isVisible()
    ).toBe(true);
  });
  it("should render loading", async () => {
    const wrapper = getWrapper();
    await wrapper.setProps({ modelValue: true, loading: true });
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should have correct HTML structure", () => {
    const wrapper = getWrapper();

    expect(wrapper.element).toMatchSnapshot();
  });

  it("should be accessible", async () => {
    const wrapper = getWrapper();
    expect(await axe(wrapper.element)).toHaveNoViolations();
  });
});
