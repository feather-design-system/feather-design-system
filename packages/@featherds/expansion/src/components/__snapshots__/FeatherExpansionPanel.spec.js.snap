// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatherExpansionPanel.vue should have correct HTML structure 1`] = `
<div
  class="feather-expansion"
>
  <div
    aria-level="4"
    role="heading"
  >
    <a
      aria-busy="false"
      aria-controls="feather-expansion-panel"
      aria-disabled="false"
      aria-expanded="false"
      class="feather-expansion-header-button"
      data-ref-id="feather-expansion-header-button"
      href="#"
      id="feather-expansion-header"
      role="button"
    >
      <span
        class="feather-expansion-header-button-text"
      >
        
        TITLE
        
      </span>
      <svg
        aria-hidden="true"
        class="feather-expansion-header-button-icon feather-icon feather-expansion-header-button-icon"
        data-v-0aa06a12=""
        focusable="false"
        height="24"
        role="img"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.71,9.29a1,1,0,0,0,0,1.41l4.59,4.59a1,1,0,0,0,1.41,0L17.3,10.7a1,1,0,0,0-1.41-1.41L12,13.17,8.12,9.29A1,1,0,0,0,6.71,9.29Z"
        />
      </svg>
    </a>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
`;

exports[`FeatherExpansionPanel.vue should not toggle expanded when disabled 1`] = `
<div
  class="feather-expansion"
>
  <div
    aria-level="4"
    role="heading"
  >
    <a
      aria-busy="false"
      aria-controls="feather-expansion-panel"
      aria-disabled="true"
      aria-expanded="false"
      class="feather-expansion-header-button disabled"
      data-ref-id="feather-expansion-header-button"
      href="#"
      id="feather-expansion-header"
      role="button"
    >
      <span
        class="feather-expansion-header-button-text"
      >
        
        TITLE
        
      </span>
      <svg
        aria-hidden="true"
        class="feather-expansion-header-button-icon disabled feather-icon feather-expansion-header-button-icon disabled"
        data-v-0aa06a12=""
        focusable="false"
        height="24"
        role="img"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.71,9.29a1,1,0,0,0,0,1.41l4.59,4.59a1,1,0,0,0,1.41,0L17.3,10.7a1,1,0,0,0-1.41-1.41L12,13.17,8.12,9.29A1,1,0,0,0,6.71,9.29Z"
        />
      </svg>
    </a>
  </div>
  <transition-stub>
    <!--v-if-->
  </transition-stub>
</div>
`;

exports[`FeatherExpansionPanel.vue should render loading 1`] = `
<div
  class="feather-expansion"
>
  <div
    aria-level="4"
    role="heading"
  >
    <a
      aria-busy="true"
      aria-controls="feather-expansion-panel"
      aria-disabled="false"
      aria-expanded="true"
      class="feather-expansion-header-button expanded"
      data-ref-id="feather-expansion-header-button"
      href="#"
      id="feather-expansion-header"
      role="button"
    >
      <span
        class="feather-expansion-header-button-text"
      >
        
        TITLE
        
      </span>
      <svg
        aria-hidden="true"
        class="feather-expansion-header-button-icon rotated feather-icon feather-expansion-header-button-icon rotated"
        data-v-0aa06a12=""
        focusable="false"
        height="24"
        role="img"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.71,9.29a1,1,0,0,0,0,1.41l4.59,4.59a1,1,0,0,0,1.41,0L17.3,10.7a1,1,0,0,0-1.41-1.41L12,13.17,8.12,9.29A1,1,0,0,0,6.71,9.29Z"
        />
      </svg>
    </a>
  </div>
  <transition-stub>
    <div
      aria-labelledby="feather-expansion-header"
      data-ref-id="feather-expansion-content"
      id="feather-expansion-panel"
      role="region"
    >
      <div
        class="panel-content"
      >
        <div
          class="spinner-container"
          data-v-137d8659=""
        >
          <svg
            class="spinner"
            data-v-137d8659=""
            viewBox="0 0 66 66"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              class="path"
              cx="33"
              cy="33"
              data-v-137d8659=""
              fill="none"
              r="30"
              stroke-linecap="round"
              stroke-width="6"
            />
          </svg>
        </div>
      </div>
    </div>
  </transition-stub>
</div>
`;
